src/Test/Goblin/Instances.hs

```diff
@@ -0,0 +1,354 @@
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TemplateHaskell       #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module Test.Goblin.Instances where
+
+import           Control.Monad (replicateM)
+import           Control.Lens
+import qualified Data.Bimap as Bimap
+import           Data.Char (chr)
+import           Data.List (splitAt)
+import qualified Data.List as List
+import qualified Data.Map as Map
+import           Data.Ratio (Ratio)
+import qualified Data.Sequence as Seq
+import qualified Data.Set as Set
+import           Data.Typeable (Typeable)
+import           Data.Word (Word8, Word64)
+import           Hedgehog (Gen)
+import qualified Hedgehog.Gen as Gen
+import           Moo.GeneticAlgorithm.Binary (bitsNeeded, decodeBinary)
+import           Numeric.Natural (Natural)
+
+import Test.Goblin.Core
+import Test.Goblin.TH
+
+
+instance GeneOps Bool where
+
+  onGene yes no = do
+    tg <- transcribeGene
+    if tg then yes else no
+
+  transcribeGenesAsInt n = do
+    (gs, xs) <- splitAt (bitsNeeded (0,n)) <$> use genes
+    genes .= xs
+    let base = n+1
+    if base == 0
+       then error "transcribeGenesAsInt: divide by zero"
+       else return $ decodeBinary (0, n) gs `mod` base
+
+--------------------------------------------------------------------------------
+-- Primitive goblins
+--------------------------------------------------------------------------------
+
+instance GeneOps a => Goblin a Bool where
+  tinker _ = addShrinks <$> onGene rummageOrConjure conjure
+  conjure = saveInBagOfTricks =<< onGene (pure True) (pure False)
+
+instance GeneOps a => Goblin a Char where
+  tinker _ = addShrinks <$> onGene rummageOrConjure conjure
+  -- TODO : this uses up 21 bits of genome, and we may not be interested in thorough
+  -- coverage of the Char space
+  conjure = saveInBagOfTricks =<< chr <$> transcribeGenesAsInt 1114111
```

Comment:
Interesting. This is quite a big departure from the initial design, in which the genome guided the _structure_ of the mutations but not the specific mutations themselves, which were somewhat random. I don't have any strong intuition as to whether this is a good/bad change, although it will massively increase the size of the genome needed, and hence slow down training.

Reply:
You're right, this is a significant change, and likely a big cause of the genome size increase. I went with this because I reasoned that it would give us higher chances of "collisions" among generated values, since they would be pulled from the genome and then inserted into the bagOfTricks, from which they could be rummaged out and reused. Since we now only have "Functor access" to the `Gen a`s inside of the Tinker monad, we can't reach inside and access the values, and then insert them into the bagOfTricks. I suppose we could try to wrangle up a mapping of `Gen a`s, so that we could associate the same generator to multiple "locations" in the larger Signal generator, which would allow us to have collisions. Does that idea make sense to you?

Reply:
My vote is to merge this PR and add work later to try to share `Gen`s, if we have an issue with genome bloat. Probably that will be an issue for Shelley.